azurelinuxagent/ga/cgroupconfigurator.py

```diff
@@ -166,94 +156,91 @@ def initialize(self):
                     except Exception as err:
                         logger.warn("Unable to delete Agent drop-in files while resetting the quotas: {0}".format(err))
 
-                # check whether cgroup monitoring is supported on the current distro
-                self._cgroups_supported = CGroupsApi.cgroups_supported()
-                if not self._cgroups_supported:
-                    logger.info("Cgroup monitoring is not supported on {0}", get_distro())
+                    log_cgroup_info("Cgroup monitoring is not supported on {0}".format(get_distro()), send_event=True)
                     return
 
                 # check that systemd is detected correctly
-                self._cgroups_api = SystemdCgroupsApi()
                 if not systemd.is_systemd():
-                    _log_cgroup_warning("systemd was not detected on {0}", get_distro())
+                    log_cgroup_warning("systemd was not detected on {0}".format(get_distro()))
                     return
 
-                _log_cgroup_info("systemd version: {0}", systemd.get_version())
+                log_cgroup_info("systemd version: {0}".format(systemd.get_version()))
+
+                # Determine which version of the Cgroup API should be used. If the correct version can't be determined,
+                # do not enable resource monitoring/enforcement.
+                try:
+                    self._cgroups_api = get_cgroup_api()
+                except CGroupsException as e:
+                    log_cgroup_warning("Unable to determine which cgroup version to use: {0}".format(ustr(e)), send_event=True)
+                    return
 
                 if not self.__check_no_legacy_cgroups():
                     return
 
                 agent_unit_name = systemd.get_agent_unit_name()
                 agent_slice = systemd.get_unit_property(agent_unit_name, "Slice")
                 if agent_slice not in (AZURE_SLICE, "system.slice"):
-                    _log_cgroup_warning("The agent is within an unexpected slice: {0}", agent_slice)
+                    log_cgroup_warning("The agent is within an unexpected slice: {0}".format(agent_slice))
                     return
 
                 self.__setup_azure_slice()
 
-                cpu_controller_root, memory_controller_root = self.__get_cgroup_controllers()
-                self._agent_cpu_cgroup_path, self._agent_memory_cgroup_path = self.__get_agent_cgroups(agent_slice,
+                cpu_controller_root, memory_controller_root = self.__get_cgroup_controller_roots()
+                self._agent_cpu_cgroup_path, self._agent_memory_cgroup_path = self.__get_agent_cgroup_paths(agent_slice,
                                                                                                        cpu_controller_root,
                                                                                                        memory_controller_root)
 
+                if self.cgroup_v2_enabled():
```

Comment:
Before we don't have this sort of implementation, we don not enable cgroups v2 based on the controller's check. Now do we need to run 193, can we move this check above?